README
nautilus-images2dxf
===========

nautilus-images2dxf

Download
--------

Download the [package][dlp] *or* checkout the source:

    $ git clone https://github.com/atareao/nautilus-images2dxf.git

Installation
------------

    $ sudo add-apt-repositoy ppa:atareao/nautilus-extensions
    $ sudo apt-get update
    $ sudo apt-get install nautilus-images2dxf


debian/changelog
nautilus-images2dxf (0.3.0-0extras16.04.0) xenail; urgency=medium

  * Convert to utm with https://github.com/Turbo87/utm

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Mon, 13 Mar 2017 16:38:14 +0100

nautilus-images2dxf (0.2.0-0extras16.04.0) xenial; urgency=medium

  * Fixed bugs

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Mon, 13 Mar 2017 15:48:16 +0100

nautilus-images2dxf (0.1.0-0extras16.04.0) xenial; urgency=medium

  * Ported to Trusty

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Mon, 13 Mar 2017 08:33:04 +0100

nautilus-images2dxf (0.1.0-0extras13.04.0) raring; urgency=low

  * Fixed a bug with gpx parsing
  * Migrated to Python3
  * Ported to Raring

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Sun, 02 Jun 2013 21:46:25 +0200

nautilus-images2dxf (0.0.1.1-1ubuntu1) precise; urgency=low

  * Fixed a bug with gpx parsing

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Sun, 26 Feb 2012 12:02:53 +0100

nautilus-images2dxf (0.0.1.0-1ubuntu1) precise; urgency=low

  * First release

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Sun, 22 Jan 2012 22:00:22 +0100


src/nautilus-images2dxf.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
# This file is part of nautilus-images2dxf
#
# Copyright (C) 2012-2016 Lorenzo Carbonell
# lorenzo.carbonell.cerezo@gmail.com
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
#
#
import gi
try:
    gi.require_version('Gtk', '3.0')
    gi.require_version('Nautilus', '3.0')
    gi.require_version('GExiv2', '0.10')
except Exception as e:
    print(e)
    exit(-1)
from gi.repository import GObject
from gi.repository import Gtk
from gi.repository import GLib
from gi.repository import GExiv2
from gi.repository import Nautilus as FileManager
import os
from threading import Thread
from urllib import unquote_plus
import mimetypes
import images2dxfutils
from images2dxfutils import sdxf
import utm

APP = '$APP$'
VERSION = '$VERSION$'

_ = str


class IdleObject(GObject.GObject):
    """
    Override GObject.GObject to always emit signals in the main thread
    by emmitting on an idle handler
    """
    def __init__(self):
        GObject.GObject.__init__(self)

    def emit(self, *args):
        GLib.idle_add(GObject.GObject.emit, self, *args)


class DoItInBackground(IdleObject, Thread):
    __gsignals__ = {
        'started': (GObject.SIGNAL_RUN_FIRST, GObject.TYPE_NONE, (int,)),
        'ended': (GObject.SIGNAL_RUN_FIRST, GObject.TYPE_NONE, (bool,)),
        'start_one': (GObject.SIGNAL_RUN_FIRST, GObject.TYPE_NONE, (str,)),
        'end_one': (GObject.SIGNAL_RUN_FIRST, GObject.TYPE_NONE, (float,)),
    }

    def __init__(self, output_filename, elements):
        IdleObject.__init__(self)
        Thread.__init__(self)
        self.elements = elements
        self.output_filename = output_filename
        self.stopit = False
        self.ok = True
        self.daemon = True
        self.process = None

    def stop(self, *args):
        self.stopit = True
        if self.process is not None:
            self.process.terminate()

    def run(self):
        total = 0
        for element in self.elements:
            total += get_duration(element)
        self.emit('started', total)
        try:
            total = 0
            drawing = images2dxfutils.sdxf.Drawing()
            for element in self.elements:
                if self.stopit is True:
                    self.ok = False
                    break
                self.emit('start_one', element)
                if os.path.exists(element) and os.path.isfile(element):
                    filename = os.path.basename(element)
                    exif = GExiv2.Metadata(element)
                    coordinates = exif.get_gps_info()
                    x, y, h, lt = utm.from_latlon(coordinates[1],
                                                  coordinates[0])
                    drawing.append(sdxf.Circle(center=(x, y, 0),
                                   radius=0.5, color=3))
                    drawing.append(sdxf.Text(filename, point=(x, y, 0)))
                self.emit('end_one', get_duration(element))
            print(1)
            drawing.saveas(self.output_filename)
            print(2)
        except Exception as e:
            print(e)
            self.ok = False
        try:
            if self.process is not None:
                self.process.terminate()
                self.process = None
        except Exception as e:
            print(e)
        print('finished')
        self.emit('ended', self.ok)


class Progreso(Gtk.Dialog, IdleObject):
    __gsignals__ = {
        'i-want-stop': (GObject.SIGNAL_RUN_FIRST, GObject.TYPE_NONE, ()),
    }

    def __init__(self, title, parent):
        Gtk.Dialog.__init__(self, title, parent,
                            Gtk.DialogFlags.MODAL |
                            Gtk.DialogFlags.DESTROY_WITH_PARENT)
        IdleObject.__init__(self)
        self.set_position(Gtk.WindowPosition.CENTER_ALWAYS)
        self.set_size_request(330, 30)
        self.set_resizable(False)
        self.connect('destroy', self.close)
        self.set_modal(True)
        vbox = Gtk.VBox(spacing=5)
        vbox.set_border_width(5)
        self.get_content_area().add(vbox)
        #
        frame1 = Gtk.Frame()
        vbox.pack_start(frame1, True, True, 0)
        table = Gtk.Table(2, 2, False)
        frame1.add(table)
        #
        self.label = Gtk.Label()
        table.attach(self.label, 0, 2, 0, 1,
                     xpadding=5,
                     ypadding=5,
                     xoptions=Gtk.AttachOptions.SHRINK,
                     yoptions=Gtk.AttachOptions.EXPAND)
        #
        self.progressbar = Gtk.ProgressBar()
        self.progressbar.set_size_request(300, 0)
        table.attach(self.progressbar, 0, 1, 1, 2,
                     xpadding=5,
                     ypadding=5,
                     xoptions=Gtk.AttachOptions.SHRINK,
                     yoptions=Gtk.AttachOptions.EXPAND)
        button_stop = Gtk.Button()
        button_stop.set_size_request(40, 40)
        button_stop.set_image(
            Gtk.Image.new_from_stock(Gtk.STOCK_STOP, Gtk.IconSize.BUTTON))
        button_stop.connect('clicked', self.on_button_stop_clicked)
        table.attach(button_stop, 1, 2, 1, 2,
                     xpadding=5,
                     ypadding=5,
                     xoptions=Gtk.AttachOptions.SHRINK)
        self.stop = False
        self.show_all()
        self.value = 0.0

    def set_max_value(self, anobject, max_value):
        self.max_value = float(max_value)

    def get_stop(self):
        return self.stop

    def on_button_stop_clicked(self, widget):
        self.stop = True
        self.emit('i-want-stop')

    def close(self, *args):
        self.destroy()

    def increase(self, anobject, value):
        self.value += float(value)
        fraction = self.value/self.max_value
        self.progressbar.set_fraction(fraction)
        if self.value >= self.max_value:
            self.hide()

    def set_element(self, anobject, element):
        self.label.set_text(_('Converting: %s') % element)


def get_files(files_in):
    files = []
    for file_in in files_in:
        print(file_in)
        file_in = unquote_plus(file_in.get_uri()[7:])
        if os.path.isfile(file_in):
            files.append(file_in)
    return files


def get_duration(file_in):
    return os.path.getsize(file_in)


def select_output_filename(window):
    output_filename = None
    dialog = Gtk.FileChooserDialog('Select dxf output file',
                                   window,
                                   Gtk.FileChooserAction.SAVE,
                                   (Gtk.STOCK_CANCEL,
                                    Gtk.ResponseType.REJECT,
                                    Gtk.STOCK_OK, Gtk.ResponseType.ACCEPT))
    dialog.set_position(Gtk.WindowPosition.CENTER_ALWAYS)
    filter = Gtk.FileFilter()
    filter.set_name('dxf file')
    filter.add_mime_type('application/dxf')
    filter.add_pattern('*.dxf')
    dialog.add_filter(filter)
    dialog.set_current_folder(os.getenv('HOME'))
    if dialog.run() == Gtk.ResponseType.ACCEPT:
        dialog.hide()
        output_filename = dialog.get_filename()
        if not output_filename.endswith('.dxf'):
            output_filename += '.dxf'
    dialog.destroy()
    return output_filename


class Images2dxfCreatorMenuProvider(GObject.GObject, FileManager.MenuProvider):
    """
    Implements the 'Replace in Filenames' extension to the File Manager\
    right-click menu
    """

    def __init__(self):
        """
        File Manager crashes if a plugin doesn't implement the __init__\
        method
        """
        mimetypes.init()
        pass

    def all_are_jpeg_files(self, items):
        for item in items:
            file_in = unquote_plus(item.get_uri()[7:])
            if not os.path.isfile(file_in):
                return False
            mimetype = mimetypes.guess_type('file://' + file_in)[0]
            if mimetype != 'image/jpeg':
                return False
        return True

    def convert(self, menu, selected, window):
        output_filename = select_output_filename(window)
        if output_filename is not None:
            files = get_files(selected)
            diib = DoItInBackground(output_filename, files)
            progreso = Progreso(_('Creating dxf'), window)
            diib.connect('started', progreso.set_max_value)
            diib.connect('start_one', progreso.set_element)
            diib.connect('end_one', progreso.increase)
            diib.connect('ended', progreso.close)
            progreso.connect('i-want-stop', diib.stop)
            diib.start()
            progreso.run()

    def get_file_items(self, window, sel_items):
        """
        Adds the 'Replace in Filenames' menu item to the File Manager\
        right-click menu, connects its 'activate' signal to the 'run'\
        method passing the selected Directory/File
        """
        if self.all_are_jpeg_files(sel_items):
            top_menuitem = FileManager.MenuItem(
                name='Images2dxfCreatorMenuProvider::Gtk-images2dxf-top',
                label=_('Create dxf from images'),
                tip=_('Tool to create a dxf from jpeg images'))
            submenu = FileManager.Menu()
            top_menuitem.set_submenu(submenu)

            sub_menuitem_00 = FileManager.MenuItem(
                name='Images2dxfCreatorMenuProvider::Gtk-images2dxf-sub-01',
                label=_('Create dxf'),
                tip=_('Tool to create a dxf from jpeg images'))
            sub_menuitem_00.connect('activate',
                                    self.convert,
                                    sel_items,
                                    window)
            submenu.append_item(sub_menuitem_00)
            sub_menuitem_01 = FileManager.MenuItem(
                name='Images2dxfCreatorMenuProvider::Gtk-images2dxf-sub-02',
                label=_('About'),
                tip=_('About'))
            sub_menuitem_01.connect('activate', self.about, window)
            submenu.append_item(sub_menuitem_01)
            #
            return top_menuitem,
        return

    def about(self, widget, window):
        ad = Gtk.AboutDialog(parent=window)
        ad.set_name(APP)
        ad.set_version(VERSION)
        ad.set_copyright('Copyrignt (c) 2017\nLorenzo Carbonell')
        ad.set_comments(APP)
        ad.set_license('''
This program is free software: you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free Software
Foundation, either version 3 of the License, or (at your option) any later
version.

This program is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program. If not, see <http://www.gnu.org/licenses/>.
''')
        ad.set_website('http://www.atareao.es')
        ad.set_website_label('http://www.atareao.es')
        ad.set_authors([
            'Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>'])
        ad.set_documenters([
            'Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>'])
        ad.set_icon_name(APP)
        ad.set_logo_icon_name(APP)
        ad.run()
        ad.destroy()


if __name__ == '__main__':
    import time
    import glob
    output_filename = select_output_filename(None)
    if output_filename is not None:
        files = glob.glob('/home/lorenzo/Escritorio/FOTOS2/*.jpg')
        print(files)
        # files = get_files(selected)
        diib = DoItInBackground(output_filename, files)
        progreso = Progreso(_('Creating dxf'), None)
        diib.connect('started', progreso.set_max_value)
        diib.connect('start_one', progreso.set_element)
        diib.connect('end_one', progreso.increase)
        diib.connect('ended', progreso.close)
        progreso.connect('i-want-stop', diib.stop)
        diib.start()
        progreso.run()
        time.sleep(5)


src/gpx2dxfutils/gpx2dxfapi.py
#!/usr/bin/python3
# -*- coding: iso-8859-1 -*-

import os
import sdxf
from utmll import LLtoUTM
from xml.dom import minidom, Node

class GPXParser:
	def __init__(self, filename):
		try:
			doc = minidom.parse(filename)
			doc.normalize()
		except:
			return # handle this properly later
		self.gpx = doc.documentElement		
	def parse(self):
		points = []
		for wpt in self.gpx.getElementsByTagName('wpt'):
			lat = float(wpt.getAttribute('lat'))
			lon = float(wpt.getAttribute('lon'))
			if len(wpt.getElementsByTagName('ele'))>0:
				ele = float(wpt.getElementsByTagName('ele')[0].firstChild.data)
			else:
				ele = 0
			E,N,Z = LLtoUTM(23,lat,lon)
			points.append((E, N, ele))
		for trkpt in self.gpx.getElementsByTagName('trkpt'):
			lat = float(trkpt.getAttribute('lat'))
			lon = float(trkpt.getAttribute('lon'))
			ele = float(trkpt.getElementsByTagName('ele')[0].firstChild.data)
			E,N,Z = LLtoUTM(23,lat,lon)
			points.append((E, N, ele))
		return points


def from_gpx_to_dxf(afile):
	if os.path.exists(afile) and os.path.isfile(afile):
		head, tail = os.path.split(afile)
		root, ext = os.path.splitext(tail)
		file_out = os.path.join(head,root+'.dxf')
		#
		gpx_file = open(afile)
		gpx_parser = GPXParser(gpx_file)
		points = gpx_parser.parse()
		drawing = sdxf.Drawing()
		if len(points)>0:
			drawing.append(sdxf.PolyLine(points=points, color=1))	
		drawing.saveas(file_out)
		return True
	return False
		

if __name__ == '__main__':
	from_gpx_to_dxf('/home/atareao/Escritorio/fells_loop.gpx')


src/images2dxfutils/images2dxfapi.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-

import os
import sdxf
from utmll import LLtoUTM
import gi
try:
    gi.require_version('GExiv2', '0.10')
except Exception as e:
    print(e)
    exit(-1)
from gi.repository import GExiv2


def from_images_to_dxf(file_out, files):
    drawing = sdxf.Drawing()
    for afile in files:
        if os.path.exists(afile) and os.path.isfile(afile):
            filename = os.path.basename(afile)
            exif = GExiv2.Metadata(afile)
            coordinates = exif.get_gps_info()
            x, y, h = LLtoUTM(23, coordinates[1], coordinates[0])
            drawing.append(sdxf.Circle(center=(x, y, 0), radius=0.5, color=3))
            drawing.append(sdxf.Text(filename, point=(x, y, 0)))
    drawing.saveas(file_out)


if __name__ == '__main__':
    exif = GExiv2.Metadata('/home/lorenzo/Escritorio/IMG_20170228_125625.jpg')
    coordinates = exif.get_gps_info()
    print(coordinates)
    x, y, h = LLtoUTM(23, coordinates[1], coordinates[0])
    print(x, y)
    files = ['/home/lorenzo/Escritorio/IMG_20170228_125625.jpg']
    from_images_to_dxf('/home/lorenzo/Escritorio/ejemplo.dxf', files)


src/gpx2dxfutils/utmll.py
#!/usr/bin/python3
# -*- coding: iso-8859-1 -*-
#
__author__='atareao'
__date__ ='$11/02/2011'
#
# A library convert from/to utm to/from longitude/latitude
#
# Copyright (C) 2011 Lorenzo Carbonell
# lorenzo.carbonell.cerezo@gmail.com
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
#
#
import math

PI = math.pi
FOURTHPI = PI / 4;
deg2rad = PI / 180;
rad2deg = 180.0 / PI;

Ellipsoid = []
#id, Ellipsoid name, Equatorial Radius, square of eccentricity	

Ellipsoid.append({'id':-1, 'name':'Placeholder','equatorial_radius':0,'eccentricity_squared':0})
Ellipsoid.append({'id': 1, 'name':'Airy','equatorial_radius':6377563,'eccentricity_squared':0.00667054})
Ellipsoid.append({'id': 2, 'name':'Australian National','equatorial_radius':6378160,'eccentricity_squared':0.006694542})
Ellipsoid.append({'id': 3, 'name':'Bessel 1841','equatorial_radius':6377397,'eccentricity_squared':0.006674372})
Ellipsoid.append({'id': 4, 'name':'Bessel 1841 (Nambia)' ,'equatorial_radius':6377484,'eccentricity_squared':0.006674372})
Ellipsoid.append({'id': 5, 'name':'Clarke 1866','equatorial_radius':6378206,'eccentricity_squared':0.006768658})
Ellipsoid.append({'id': 6, 'name':'Clarke 1880','equatorial_radius':6378249,'eccentricity_squared':0.006803511})
Ellipsoid.append({'id': 7, 'name':'Everest','equatorial_radius':6377276,'eccentricity_squared':0.006637847})
Ellipsoid.append({'id': 8, 'name':'Fischer 1960 (Mercury)' ,'equatorial_radius':6378166,'eccentricity_squared':0.006693422})
Ellipsoid.append({'id': 9, 'name':'Fischer 1968','equatorial_radius':6378150,'eccentricity_squared':0.006693422})
Ellipsoid.append({'id': 10, 'name':'GRS 1967','equatorial_radius':6378160,'eccentricity_squared':0.006694605})
Ellipsoid.append({'id': 11, 'name':'GRS 1980','equatorial_radius':6378137,'eccentricity_squared':0.00669438})
Ellipsoid.append({'id': 12, 'name':'Helmert 1906','equatorial_radius':6378200,'eccentricity_squared':0.006693422})
Ellipsoid.append({'id': 13, 'name':'Hough','equatorial_radius':6378270,'eccentricity_squared':0.00672267})
Ellipsoid.append({'id': 14, 'name':'International','equatorial_radius':6378388,'eccentricity_squared':0.00672267})
Ellipsoid.append({'id': 15, 'name':'Krassovsky','equatorial_radius':6378245,'eccentricity_squared':0.006693422})
Ellipsoid.append({'id': 16, 'name':'Modified Airy','equatorial_radius':6377340,'eccentricity_squared':0.00667054})
Ellipsoid.append({'id': 17, 'name':'Modified Everest','equatorial_radius':6377304,'eccentricity_squared':0.006637847})
Ellipsoid.append({'id': 18, 'name':'Modified Fischer 1960','equatorial_radius':6378155,'eccentricity_squared':0.006693422})
Ellipsoid.append({'id': 19, 'name':'South American 1969','equatorial_radius':6378160,'eccentricity_squared':0.006694542})
Ellipsoid.append({'id': 20, 'name':'WGS 60','equatorial_radius':6378165,'eccentricity_squared':0.006693422})
Ellipsoid.append({'id': 21, 'name':'WGS 66','equatorial_radius':6378145,'eccentricity_squared':0.006694542})
Ellipsoid.append({'id': 22, 'name':'WGS-72','equatorial_radius':6378135,'eccentricity_squared':0.006694318})
Ellipsoid.append({'id': 23, 'name':'WGS-84','equatorial_radius':6378137,'eccentricity_squared':0.00669438})



def LLtoUTM(ellipsoid, lat, lon):
	a = Ellipsoid[ellipsoid]['equatorial_radius']
	eccSquared = Ellipsoid[ellipsoid]['eccentricity_squared']
	k0 = 0.9996;
	LongOrigin = 0.0
	eccPrimeSquared = 0.0
	N = 0.0
	T = 0.0
	C = 0.0
	A = 0.0
	M = 0.0
	LongTemp = (lon+180)-int((lon+180)/360)*360-180#; // -180.00 .. 179.9;
	LatRad = lat*deg2rad
	LongRad = LongTemp*deg2rad
	ZoneNumber = Huso(lat,lon)
	LongOrigin = (ZoneNumber - 1)*6 - 180 + 3#+3 puts origin in middle of zone
	LongOriginRad = LongOrigin * deg2rad
	#compute the UTM Zone from the latitude and longitude
	#sprintf(UTMZone, "%d%c", ZoneNumber, UTMLetterDesignator(Lat));
	eccPrimeSquared = (eccSquared)/(1-eccSquared)
	N = a/math.sqrt(1-eccSquared*math.sin(LatRad)*math.sin(LatRad))
	T = math.tan(LatRad)*math.tan(LatRad)
	C = eccPrimeSquared*math.cos(LatRad)*math.cos(LatRad)
	A = math.cos(LatRad)*(LongRad-LongOriginRad)
	M = a*((1-eccSquared/4-3*math.pow(eccSquared,2)/64-5*math.pow(eccSquared,3)/256)*LatRad-(3*eccSquared/8+3*math.pow(eccSquared,2)/32+45*math.pow(eccSquared,3)/1024)*math.sin(2*LatRad)+(15*math.pow(eccSquared,2)/256+45*math.pow(eccSquared,3)/1024)*math.sin(4*LatRad)-(35*math.pow(eccSquared,3)/3072)*math.sin(6*LatRad))
	UTMEasting = k0*N*(A+(1-T+C)*math.pow(A,3)/6 + (5-18*T+math.pow(T,2)+72*C-58*eccPrimeSquared)*math.pow(A,5)/120)+500000.0
	UTMNorthing = k0*(M+N*math.tan(LatRad)*(math.pow(A,2)/2+(5-T+9*C+4*math.pow(C,2))*math.pow(A,4)/24 + (61-58*T+math.pow(T,2)+600*C-330*eccPrimeSquared)*math.pow(A,6)/720))
	if lat < 0:
		UTMNorthing += 10000000.0#10000000 meter offset for southern hemisphere
	return UTMEasting, UTMNorthing, str(ZoneNumber)+Banda(lat)
	
	
def Huso(latitud,longitud):
	LongTemp = (longitud+180)-int((longitud+180)/360)*360-180#; // -180.00 .. 179.9;
	huso = int((LongTemp + 180)/6) + 1
	if latitud >= 56.0 and latitud < 64.0 and LongTemp >= 3.0 and LongTemp < 12.0:
		huso = 32
	#Special zones for Svalbard
	if latitud >= 72.0 and latitud < 84.0:
		if LongTemp >= 0.0  and LongTemp <  9.0:
			huso = 31
		elif LongTemp >= 9.0 and LongTemp < 21.0:
			huso = 33
		elif LongTemp >= 21.0 and LongTemp < 33.0:
			huso = 35
		elif LongTemp >= 33.0 and LongTemp < 42.0:
			huso = 37
	return huso

	
def Banda(lat):
	letter = 'Z'
	if 84 >= lat and lat >= 72:
		letter = 'X'
	elif 72 > lat and lat >= 64:
		letter = 'W'
	elif 64 > lat and lat >= 56:
		letter = 'V'
	elif 56 > lat and lat >= 48:
		letter = 'U'
	elif 48 > lat and lat >= 40:
		letter = 'T'
	elif 40 > lat and lat >= 32:
		letter = 'S'
	elif 32 > lat and lat >= 24:
		letter = 'R'
	elif 24 > lat and lat >= 16:
		letter = 'Q'
	elif 16 > lat and lat >= 8:
		letter = 'P'
	elif 8 > lat and lat >= 0:
		letter = 'N'
	elif 0 > lat and lat >= -8:
		letter = 'M'
	elif-8> lat and lat >= -16:
		letter = 'L'
	elif-16 > lat and lat >= -24:
		letter = 'K'
	elif-24 > lat and lat >= -32:
		letter = 'J'
	elif-32 > lat and lat >= -40:
		letter = 'H'
	elif-40 > lat and lat >= -48:
		letter = 'G'
	elif-48 > lat and lat >= -56:
		letter = 'F'
	elif-56 > lat and lat >= -64:
		letter = 'E'
	elif-64 > lat and lat >= -72:
		letter = 'D'
	elif-72 > lat and lat >= -80:
		letter = 'C'
	return letter
	
#print LLtoUTM(23,39.36,-0.41)
#print UTMLetter(39.36)

def UTMtoLL(ellipsoid, UTMEasting, UTMNorthing, UTMZone):
	k0 = 0.9996
	a = Ellipsoid[ellipsoid]['equatorial_radius']
	eccSquared = Ellipsoid[ellipsoid]['eccentricity_squared']
	eccPrimeSquared = 0.0
	e1 = (1-math.sqrt(1-eccSquared))/(1+math.sqrt(1-eccSquared))
	N1 = 0.0
	T1 = 0.0
	C1 = 0.0
	R1 = 0.0
	D = 0.0
	M = 0.0
	LongOrigin = 0.0
	mu = 0.0
	phi1 = 0.0
	phi1Rad = 0.0
	x = 0.0
	y = 0.0
	ZoneNumber = 0
	NorthernHemisphere = 1 # 1 for northern hemispher, 0 for southern
	x = UTMEasting - 500000.0 #remove 500,000 meter offset for longitude
	y = UTMNorthing
	ZoneNumber=int(UTMZone[0:(len(UTMZone)-1)])
	ZoneLetter=UTMZone[(len(UTMZone)-1):]
	if (ord(ZoneLetter.upper()) - ord('N')) >= 0:
		NorthernHemisphere = 1 # point is in northern hemisphere
	else:
		NorthernHemisphere = 0 # point is in southern hemisphere
		y -= 10000000.0 # remove 10,000,000 meter offset used for southern hemisphere
	print(ZoneNumber, ZoneLetter, NorthernHemisphere)
	LongOrigin = (ZoneNumber - 1)*6 - 180 + 3#;  //+3 puts origin in middle of zone
	eccPrimeSquared = (eccSquared)/(1-eccSquared)
	M = y / k0
	mu = M/(a*(1-eccSquared/4-3*eccSquared*eccSquared/64-5*eccSquared*eccSquared*eccSquared/256))
	phi1Rad = mu+(3*e1/2-27*math.pow(e1,3)/32)*math.sin(2*mu)+(21*math.pow(e1,2)/16-55*math.pow(e1,4)/32)*math.sin(4*mu)+(151*math.pow(e1,3)/96)*math.sin(6*mu)
	phi1 = phi1Rad*rad2deg
	N1 = a/math.sqrt(1-eccSquared*math.sin(phi1Rad)*math.sin(phi1Rad))
	T1 = math.tan(phi1Rad)*math.tan(phi1Rad)
	C1 = eccPrimeSquared*math.cos(phi1Rad)*math.cos(phi1Rad)
	R1 = a*(1-eccSquared)/math.pow(1-eccSquared*math.sin(phi1Rad)*math.sin(phi1Rad), 1.5)
	D = x/(N1*k0);
	Lat = phi1Rad - (N1*math.tan(phi1Rad)/R1)*(math.pow(D,2)/2-(5+3*T1+10*C1-4*math.pow(C1,2)-9*eccPrimeSquared)*math.pow(D,4)/24+(61+90*T1+298*C1+45*math.pow(T1,2)-252*eccPrimeSquared-3*math.pow(C1,2))*math.pow(D,6)/720)
	Lat = Lat * rad2deg
	Long = (D-(1+2*T1+C1)*math.pow(D,3)/6+(5-2*C1+28*T1-3*math.pow(C1,2)+8*eccPrimeSquared+24*T1*T1)*math.pow(D,5)/120)/math.cos(phi1Rad)
	Long = LongOrigin + Long * rad2deg
	return Lat,Long
if __name__ == '__main__':
	E,N,Z = LLtoUTM (23,0,40)
	print(E,N,Z)
	exit(0)
